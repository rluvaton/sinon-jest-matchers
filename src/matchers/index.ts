export { sinonToBeCalled } from './sinonToBeCalled';
export { sinonToBeCalledTimes } from './sinonToBeCalledTimes';
export { sinonToBeCalledWith } from './sinonToBeCalledWith';
export { sinonLastCalledWith } from './sinonLastCalledWith';
export { sinonLastReturnedWith } from './sinonLastReturnedWith';
export { sinonNthCalledWith } from './sinonNthCalledWith';
export { sinonToReturn } from './sinonToReturn';
export { sinonToReturnTimes } from './sinonToReturnTimes';
export { sinonToReturnWith } from './sinonToReturnWith';
