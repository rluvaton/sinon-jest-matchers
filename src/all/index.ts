import * as matchers from '../matchers';

// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore
const jestExpect = global.expect;

if (jestExpect !== undefined) {
  jestExpect.extend(matchers);
} else {
  throw new Error(
    "Unable to find Jest's global expect. " +
      'Please check you have added sinon-jest-matchers correctly to your jest configuration. ' +
      'See https://github.com/rluvaton/sinon-jest-matchers#setup for help.',
  );
}
