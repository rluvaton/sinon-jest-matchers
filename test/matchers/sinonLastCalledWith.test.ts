import * as sinon from 'sinon';
import * as matchers from '../../src';
import * as Immutable from 'immutable';

expect.extend(matchers);

expect.extend({
  optionalFn(fn?: unknown) {
    const pass = fn === undefined || typeof fn === 'function';
    return { message: () => 'expect either a function or undefined', pass };
  },
});

describe('sinonLastCalledWith', () => {
  test('works only on spies or sinon.spy', () => {
    const fn = function fn() {};

    expect(() => expect(fn).sinonLastCalledWith()).toThrowErrorMatchingSnapshot();
  });

  test('works when not called', () => {
    const fn = sinon.spy();
    expect(fn).not.sinonLastCalledWith('foo', 'bar');

    expect(() => expect(fn).sinonLastCalledWith('foo', 'bar')).toThrowErrorMatchingSnapshot();
  });

  test('works with no arguments', () => {
    const fn = sinon.spy();
    fn();
    expect(fn).sinonLastCalledWith();
  });

  test("works with arguments that don't match", () => {
    const fn = sinon.spy();
    fn('foo', 'bar1');

    expect(fn).not.sinonLastCalledWith('foo', 'bar');

    expect(() => expect(fn).sinonLastCalledWith('foo', 'bar')).toThrowErrorMatchingSnapshot();
  });

  test("works with arguments that don't match in number of arguments", () => {
    const fn = sinon.spy();
    fn('foo', 'bar', 'plop');

    expect(fn).not.sinonLastCalledWith('foo', 'bar');

    expect(() => expect(fn).sinonLastCalledWith('foo', 'bar')).toThrowErrorMatchingSnapshot();
  });

  test("works with arguments that don't match with matchers", () => {
    const fn = sinon.spy();
    fn('foo', 'bar');

    expect(fn).not.sinonLastCalledWith(expect.any(String), expect.any(Number));

    expect(() => expect(fn).sinonLastCalledWith(expect.any(String), expect.any(Number))).toThrowErrorMatchingSnapshot();
  });

  test("works with arguments that don't match with matchers even when argument is undefined", () => {
    const fn = sinon.spy();
    fn('foo', undefined);

    expect(fn).not.sinonLastCalledWith('foo', expect.any(String));

    expect(() => expect(fn).sinonLastCalledWith('foo', expect.any(String))).toThrowErrorMatchingSnapshot();
  });

  // Test fails, will ignore for now until it will because an actual problem
  test.skip("works with arguments that don't match in size even if one is an optional matcher", () => {
    const fn = sinon.spy();
    fn('foo');

    expect(fn).not.sinonLastCalledWith('foo', (expect as any).optionalFn());
    expect(() => expect(fn).sinonLastCalledWith('foo', (expect as any).optionalFn())).toThrowErrorMatchingSnapshot();
  });

  test('works with arguments that match', () => {
    const fn = sinon.spy();
    fn('foo', 'bar');

    expect(fn).sinonLastCalledWith('foo', 'bar');

    expect(() => expect(fn).not.sinonLastCalledWith('foo', 'bar')).toThrowErrorMatchingSnapshot();
  });

  test('works with arguments that match with matchers', () => {
    const fn = sinon.spy();
    fn('foo', 'bar');

    expect(fn).sinonLastCalledWith(expect.any(String), expect.any(String));

    expect(() =>
      expect(fn).not.sinonLastCalledWith(expect.any(String), expect.any(String)),
    ).toThrowErrorMatchingSnapshot();
  });

  // Test fails, will ignore for now until it will because an actual problem
  test.skip('works with trailing undefined arguments', () => {
    const fn = sinon.spy();
    fn('foo', undefined);

    expect(() => expect(fn).sinonLastCalledWith('foo')).toThrowErrorMatchingSnapshot();
  });

  test('works with trailing undefined arguments if requested by the match query', () => {
    const fn = sinon.spy();
    fn('foo', undefined);

    expect(fn).sinonLastCalledWith('foo', undefined);
    expect(() => expect(fn).not.sinonLastCalledWith('foo', undefined)).toThrowErrorMatchingSnapshot();
  });

  test('works with trailing undefined arguments when explicitly requested as optional by matcher', () => {
    const fn = sinon.spy();
    fn('foo', undefined);

    expect(fn).sinonLastCalledWith('foo', (expect as any).optionalFn());
    expect(() =>
      expect(fn).not.sinonLastCalledWith('foo', (expect as any).optionalFn()),
    ).toThrowErrorMatchingSnapshot();
  });

  test('works with Map', () => {
    const fn = sinon.spy();

    const m1 = new Map([
      [1, 2],
      [2, 1],
    ]);
    const m2 = new Map([
      [1, 2],
      [2, 1],
    ]);
    const m3 = new Map([
      ['a', 'b'],
      ['b', 'a'],
    ]);

    fn(m1);

    expect(fn).sinonLastCalledWith(m2);
    expect(fn).not.sinonLastCalledWith(m3);

    expect(() => expect(fn).not.sinonLastCalledWith(m2)).toThrowErrorMatchingSnapshot();
    expect(() => expect(fn).sinonLastCalledWith(m3)).toThrowErrorMatchingSnapshot();
  });

  test('works with Set', () => {
    const fn = sinon.spy();

    const s1 = new Set([1, 2]);
    const s2 = new Set([1, 2]);
    const s3 = new Set([3, 4]);

    fn(s1);

    expect(fn).sinonLastCalledWith(s2);
    expect(fn).not.sinonLastCalledWith(s3);

    expect(() => expect(fn).not.sinonLastCalledWith(s2)).toThrowErrorMatchingSnapshot();
    expect(() => expect(fn).sinonLastCalledWith(s3)).toThrowErrorMatchingSnapshot();
  });

  test('works with Immutable.js objects', () => {
    const fn = sinon.spy();
    const directlyCreated = Immutable.Map([['a', { b: 'c' }]]);
    const indirectlyCreated = Immutable.Map().set('a', { b: 'c' });
    fn(directlyCreated, indirectlyCreated);

    expect(fn).sinonLastCalledWith(indirectlyCreated, directlyCreated);

    expect(() => expect(fn).not.sinonLastCalledWith(indirectlyCreated, directlyCreated)).toThrowErrorMatchingSnapshot();
  });

  test('works with many arguments', () => {
    const fn = sinon.spy();
    fn('foo1', 'bar');
    fn('foo', 'bar1');
    fn('foo', 'bar');

    expect(fn).sinonLastCalledWith('foo', 'bar');

    expect(() => expect(fn).not.sinonLastCalledWith('foo', 'bar')).toThrowErrorMatchingSnapshot();
  });

  test("works with many arguments that don't match", () => {
    const fn = sinon.spy();
    fn('foo', 'bar1');
    fn('foo', 'bar2');
    fn('foo', 'bar3');

    expect(fn).not.sinonLastCalledWith('foo', 'bar');

    expect(() => expect(fn).sinonLastCalledWith('foo', 'bar')).toThrowErrorMatchingSnapshot();
  });

  test('includes the custom mock name in the error message', () => {
    const fn = sinon.spy();
    fn.named('named-mock');
    fn('foo', 'bar');

    expect(fn).sinonLastCalledWith('foo', 'bar');

    expect(() => expect(fn).not.sinonLastCalledWith('foo', 'bar')).toThrowErrorMatchingSnapshot();
  });
});
