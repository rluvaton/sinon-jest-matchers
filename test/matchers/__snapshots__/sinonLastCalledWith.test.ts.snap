// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sinonLastCalledWith includes the custom mock name in the error message 1`] = `
"[2mexpect([22m[31mnamed-mock[39m[2m).[22mnot[2m.[22msinonLastCalledWith[2m([22m[32m...expected[39m[2m)[22m

Expected: not [32m\\"foo\\"[39m, [32m\\"bar\\"[39m

Number of calls: [31m1[39m"
`;

exports[`sinonLastCalledWith works only on spies or sinon.spy 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).[22msinonLastCalledWith[2m([22m[32m...expected[39m[2m)[22m

[1mMatcher error[22m: [31mreceived[39m value must be a stub or spy function

Received has type:  function
Received has value: [31m[Function fn][39m"
`;

exports[`sinonLastCalledWith works when not called 1`] = `
"[2mexpect.[22msinonLastCalledWith[2m([22m[32m...expected[39m[2m)[22m

Expected: [32m\\"foo\\"[39m, [32m\\"bar\\"[39m

Number of calls: [31m0[39m"
`;

exports[`sinonLastCalledWith works with Immutable.js objects 1`] = `
"[2mexpect.[22mnot[2m.[22msinonLastCalledWith[2m([22m[32m...expected[39m[2m)[22m

Expected: not [32mImmutable.Map {\\"a\\": {\\"b\\": \\"c\\"}}[39m, [32mImmutable.Map {\\"a\\": {\\"b\\": \\"c\\"}}[39m

Number of calls: [31m1[39m"
`;

exports[`sinonLastCalledWith works with Map 1`] = `
"[2mexpect.[22mnot[2m.[22msinonLastCalledWith[2m([22m[32m...expected[39m[2m)[22m

Expected: not [32mMap {1 => 2, 2 => 1}[39m

Number of calls: [31m1[39m"
`;

exports[`sinonLastCalledWith works with Map 2`] = `
"[2mexpect.[22msinonLastCalledWith[2m([22m[32m...expected[39m[2m)[22m

[32m- Expected[39m
[31m+ Received[39m

[2m  Map {[22m
[32m-   \\"a\\" => \\"b\\",[39m
[32m-   \\"b\\" => \\"a\\",[39m
[31m+   1 => 2,[39m
[31m+   2 => 1,[39m
[2m  }[22m,

Number of calls: [31m1[39m"
`;

exports[`sinonLastCalledWith works with Set 1`] = `
"[2mexpect.[22mnot[2m.[22msinonLastCalledWith[2m([22m[32m...expected[39m[2m)[22m

Expected: not [32mSet {1, 2}[39m

Number of calls: [31m1[39m"
`;

exports[`sinonLastCalledWith works with Set 2`] = `
"[2mexpect.[22msinonLastCalledWith[2m([22m[32m...expected[39m[2m)[22m

[32m- Expected[39m
[31m+ Received[39m

[2m  Set {[22m
[32m-   3,[39m
[32m-   4,[39m
[31m+   1,[39m
[31m+   2,[39m
[2m  }[22m,

Number of calls: [31m1[39m"
`;

exports[`sinonLastCalledWith works with arguments that don't match 1`] = `
"[2mexpect.[22msinonLastCalledWith[2m([22m[32m...expected[39m[2m)[22m

Expected: [32m\\"foo\\"[39m, [32m\\"bar\\"[39m
Received: [2m\\"foo\\"[22m, [31m\\"bar1\\"[39m

Number of calls: [31m1[39m"
`;

exports[`sinonLastCalledWith works with arguments that don't match in number of arguments 1`] = `
"[2mexpect.[22msinonLastCalledWith[2m([22m[32m...expected[39m[2m)[22m

Expected: [32m\\"foo\\"[39m, [32m\\"bar\\"[39m
Received: [2m\\"foo\\"[22m, [2m\\"bar\\"[22m, [31m\\"plop\\"[39m

Number of calls: [31m1[39m"
`;

exports[`sinonLastCalledWith works with arguments that don't match with matchers 1`] = `
"[2mexpect.[22msinonLastCalledWith[2m([22m[32m...expected[39m[2m)[22m

Expected: [32mAny<String>[39m, [32mAny<Number>[39m
Received: [2m\\"foo\\"[22m, [31m\\"bar\\"[39m

Number of calls: [31m1[39m"
`;

exports[`sinonLastCalledWith works with arguments that don't match with matchers even when argument is undefined 1`] = `
"[2mexpect.[22msinonLastCalledWith[2m([22m[32m...expected[39m[2m)[22m

Expected: [32m\\"foo\\"[39m, [32mAny<String>[39m
Received: [2m\\"foo\\"[22m, [31mundefined[39m

Number of calls: [31m1[39m"
`;

exports[`sinonLastCalledWith works with arguments that match 1`] = `
"[2mexpect.[22mnot[2m.[22msinonLastCalledWith[2m([22m[32m...expected[39m[2m)[22m

Expected: not [32m\\"foo\\"[39m, [32m\\"bar\\"[39m

Number of calls: [31m1[39m"
`;

exports[`sinonLastCalledWith works with arguments that match with matchers 1`] = `
"[2mexpect.[22mnot[2m.[22msinonLastCalledWith[2m([22m[32m...expected[39m[2m)[22m

Expected: not [32mAny<String>[39m, [32mAny<String>[39m
Received:     [31m0[39m, [31m[\\"foo\\", \\"bar\\"][39m

Number of calls: [31m1[39m"
`;

exports[`sinonLastCalledWith works with many arguments 1`] = `
"[2mexpect.[22mnot[2m.[22msinonLastCalledWith[2m([22m[32m...expected[39m[2m)[22m

Expected: not [32m\\"foo\\"[39m, [32m\\"bar\\"[39m
Received
       2:     [2m\\"foo\\"[22m, [31m\\"bar1\\"[39m
->     3:     [2m\\"foo\\"[22m, [2m\\"bar\\"[22m

Number of calls: [31m3[39m"
`;

exports[`sinonLastCalledWith works with many arguments that don't match 1`] = `
"[2mexpect.[22msinonLastCalledWith[2m([22m[32m...expected[39m[2m)[22m

Expected: [32m\\"foo\\"[39m, [32m\\"bar\\"[39m
Received
       2: [2m\\"foo\\"[22m, [31m\\"bar2\\"[39m
->     3: [2m\\"foo\\"[22m, [31m\\"bar3\\"[39m

Number of calls: [31m3[39m"
`;

exports[`sinonLastCalledWith works with trailing undefined arguments if requested by the match query 1`] = `
"[2mexpect.[22mnot[2m.[22msinonLastCalledWith[2m([22m[32m...expected[39m[2m)[22m

Expected: not [32m\\"foo\\"[39m, [32mundefined[39m

Number of calls: [31m1[39m"
`;

exports[`sinonLastCalledWith works with trailing undefined arguments when explicitly requested as optional by matcher 1`] = `
"[2mexpect.[22mnot[2m.[22msinonLastCalledWith[2m([22m[32m...expected[39m[2m)[22m

Expected: not [32m\\"foo\\"[39m, [32moptionalFn<>[39m
Received:     [31m0[39m, [31m[\\"foo\\", undefined][39m

Number of calls: [31m1[39m"
`;
